sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/DataReaderFactory.java

```diff
@@ -52,10 +56,46 @@
   }
 
   /**
-   * Returns a data reader to do the actual reading work.
+   * The output data format of this factory's data reader. Spark will invoke the corresponding
+   * create data reader method w.r.t. the return value of this method:
+   * <ul>
+   *   <li>{@link DataFormat#ROW}: {@link #createRowDataReader()}</li>
+   *   <li>{@link DataFormat#UNSAFE_ROW}: {@link #createUnsafeRowDataReader()}</li>
+   *   <li>@{@link DataFormat#COLUMNAR_BATCH}: {@link #createColumnarBatchDataReader()}</li>
+   * </ul>
+   */
+  DataFormat dataFormat();
```

Comment:
If the data format is determined when the factory is created, then I don't see why it is necessary to change the API. This just makes it more confusing.